More useful error handling

When trying to run my (obviously broken) script, I get the following error:

```
panic: regexp: Compile(`*\.* @@.*`): error parsing regexp: missing argument to repetition operator: `*`

goroutine 10 [running]:
panic(0x6e2940, 0xc82028ea80)
        /home/travis/.gimme/versions/go1.6.1.linux.amd64/src/runtime/panic.go:464 +0x3e6
regexp.MustCompile(0xc8200c7dc1, 0x9, 0x7d9648)
        /home/travis/.gimme/versions/go1.6.1.linux.amd64/src/regexp/regexp.go:232 +0x16f
github.com/aelsabbahy/goss/resource.newRegexPattern(0xc8200c7dc0, 0xb, 0x7d9a48)
        /home/travis/gopath/src/github.com/aelsabbahy/goss/resource/validate.go:169 +0x1e4
github.com/aelsabbahy/goss/resource.sliceToPatterns(0xc8200f33a0, 0x1, 0x1, 0x0, 0x0, 0x0)
        /home/travis/gopath/src/github.com/aelsabbahy/goss/resource/validate.go:190 +0x428
github.com/aelsabbahy/goss/resource.ValidateContains(0x7fdf2e39adb8, 0xc820072f00, 0x7e4df0, 0x8, 0xc8200f33a0, 0x1, 0x1, 0xc82012be20, 0x0, 0x0, ...)
        /home/travis/gopath/src/github.com/aelsabbahy/goss/resource/validate.go:228 +0x46f
github.com/aelsabbahy/goss/resource.(*File).Validate(0xc820072f00, 0xc820072640, 0x0, 0x0, 0x0)
        /home/travis/gopath/src/github.com/aelsabbahy/goss/resource/file.go:55 +0xb7e
github.com/aelsabbahy/goss.Validate.func2(0xc820101080, 0xc8200eb020, 0xc8200eafc0, 0xc820072640)
        /home/travis/gopath/src/github.com/aelsabbahy/goss/validate.go:69 +0xb3
created by github.com/aelsabbahy/goss.Validate
        /home/travis/gopath/src/github.com/aelsabbahy/goss/validate.go:72 +0x483
```

Although I recognise the error is somewhere in my code and not something in goss, but it would be helpful if it could point out a line number or something of the yaml.

--- a/integration-tests/goss/centos6/goss-expected.json
+++ b/integration-tests/goss/centos6/goss-expected.json
@@ -97,7 +97,7 @@
             "exit-status": 0,
             "stdout": [
                 "Server version: Apache/2.2.15 (Unix)",
-                "Server built:   Jul 24 2015 11:52:28"
+                "Server built:   May 11 2016 19:28:33"
             ],
             "stderr": [],
             "timeout": 10000

--- a/integration-tests/goss/centos6/goss.json
+++ b/integration-tests/goss/centos6/goss.json
@@ -39,7 +39,7 @@
             "exit-status": 0,
             "stdout": [
                 "Server version: Apache/2.2.15 (Unix)",
-                "Server built:   Jul 24 2015 11:52:28"
+                "Server built:   May 11 2016 19:28:33"
             ],
             "stderr": []
         },

--- a/integration-tests/goss/goss-shared.json
+++ b/integration-tests/goss/goss-shared.json
@@ -35,7 +35,7 @@
         },
         "tcp://google.com:443": {
             "reachable": true,
-            "timeout": 500
+            "timeout": 2000
         }
     },
     "port": {

--- a/resource/validate.go
+++ b/resource/validate.go
@@ -144,7 +144,7 @@ type regexPattern struct {
 	inverse bool
 }
 
-func newRegexPattern(str string) *regexPattern {
+func newRegexPattern(str string) (*regexPattern, error) {
 	var inverse bool
 	cleanStr := str
 	if strings.HasPrefix(str, "!") {
@@ -165,14 +165,14 @@ func newRegexPattern(str string) *regexPattern {
 			break
 		}
 	}
-	// fixme, don't use MustCompile
-	re := regexp.MustCompile(cleanStr)
+
+	re, err := regexp.Compile(cleanStr)
 
 	return &regexPattern{
 		pattern: str,
 		re:      re,
 		inverse: inverse,
-	}
+	}, err
 
 }
 
@@ -183,16 +183,20 @@ func (re *regexPattern) Match(str string) bool {
 func (re *regexPattern) Pattern() string { return re.pattern }
 func (re *regexPattern) Inverse() bool   { return re.inverse }
 
-func sliceToPatterns(slice []string) []patternMatcher {
+func sliceToPatterns(slice []string) ([]patternMatcher, error) {
 	var patterns []patternMatcher
 	for _, s := range slice {
 		if (strings.HasPrefix(s, "/") || strings.HasPrefix(s, "!/")) && strings.HasSuffix(s, "/") {
-			patterns = append(patterns, newRegexPattern(s))
+			pat, err := newRegexPattern(s)
+			if err != nil {
+				return nil, err
+			}
+			patterns = append(patterns, pat)
 		} else {
 			patterns = append(patterns, newStringPattern(s))
 		}
 	}
-	return patterns
+	return patterns, nil
 }
 
 func patternsToSlice(patterns []patternMatcher) []string {
@@ -210,7 +214,13 @@ func ValidateContains(res ResourceRead, property string, expectedValues []string
 	typ := reflect.TypeOf(res)
 	typs := strings.Split(typ.String(), ".")[1]
 	startTime := time.Now()
-	fh, err := method()
+	var err error
+	var fh io.Reader
+	var notfound []patternMatcher
+	fh, err = method()
+	if err == nil {
+		notfound, err = sliceToPatterns(expectedValues)
+	}
 	if err != nil {
 		return TestResult{
 			Successful:   false,
@@ -225,7 +235,6 @@ func ValidateContains(res ResourceRead, property string, expectedValues []string
 		}
 	}
 	scanner := bufio.NewScanner(fh)
-	notfound := sliceToPatterns(expectedValues)
 	var found []patternMatcher
 	for scanner.Scan() {
 		line := scanner.Text()

--- a/resource/validate_test.go
+++ b/resource/validate_test.go
@@ -104,3 +104,14 @@ func TestValidateContainsErr(t *testing.T) {
 		}
 	}
 }
+
+func TestValidateContainsBadRegexErr(t *testing.T) {
+	inFunc := func() (io.Reader, error) {
+		reader := strings.NewReader("dummy")
+		return reader, nil
+	}
+	got := ValidateContains(&FakeResource{""}, "", []string{"/*\\.* @@.*/"}, inFunc)
+	if got.Err == nil {
+		t.Errorf("Expected bad regex to raise error, got nil")
+	}
+}

